Add DateTimeOffset parameter check and some tests

stevenaw opened the pull request:
Fixes #148 

This PR adds support for detecting valid string -> DateTimeOffset conversions in TestCase.

stevenaw commented on src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs:
```diff
@@ -64,6 +64,11 @@ internal static bool CanAssignTo(this AttributeArgumentSyntax @this, ITypeSymbol
+                    else if (argumentType.SpecialType == SpecialType.System_String &&
+                        model.Compilation.GetTypeByMetadataName(typeof(DateTimeOffset).FullName).IsAssignableFrom(targetType))
+                    {
+                        canConvert = DateTimeOffset.TryParse(argumentValue as string, out _);
```
Note to reviewer: I used `TryParse` here, but I think the underlying NUnit implementation uses `TypeConverter.CanConvertFrom`. Please let me know if you'd prefer I update it.

Reference code: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/ParamAttributeTypeConversions.cs#L138-L143

mikkelbu replied:
Actually, I don't know the exact differences between `DateTimeOffset.TryParse` and `TypeConverter.CanConvertFrom`, so I'm unsure about this. But I think it will be the safest to have the same implementation, if it is not too cumbersome. Also I think we should use `CultureInfo.InvariantCulture`.

stevenaw replied:
Thanks for the feedback. I think I have this working now using `TypeConverter` + `CultureInfo.InvariantCulture` :)

stevenaw commented on src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs:
```diff
@@ -51,6 +51,29 @@ public void VerifySupportedDiagnostics()
         }
 
+
+        private static IEnumerable<TestCaseData> SpecialConversions
```
Note to reviewer: These have been duplicated across test files because I couldn't find a good place to centralize them. Please let me know if you have any suggestions, or think its ok as-is.

mikkelbu replied:
I think it is fine as it is, but perhaps also add a testcase with a time and offset for `DateTimeOffset`.

mikkelbu commented on src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs:
```diff
@@ -100,6 +110,30 @@ public sealed class AnalyzeWhenArgumentIsCorrect
+        }
+        
+        [Test]
+        public void AnalyzeWhenArgumentIsVerySpecialConversion()
```
This test looks the same as the `TestCaseData("2019-10-10", typeof(DateTimeOffset))` case above?

stevenaw replied:
Oh you're right, thanks for spotting this! I'd been rushing to get this PR up before my laptop battery died, and I must've missed cleaning up this earlier attempt at a test.

mikkelbu commented on src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs:
```diff
@@ -15,6 +15,16 @@ namespace NUnit.Analyzers.Tests.TestCaseUsage
+            {
+                yield return new TestCaseData("2019-10-10", typeof(DateTime));
+                yield return new TestCaseData("23:59:59", typeof(TimeSpan));
+                yield return new TestCaseData("2019-10-10", typeof(DateTimeOffset));
```
Can you also provide a testcase where we have time and offset for the `DateTimeOffset`?

stevenaw commented:
Thanks for the review @mikkelbu 
I think I've covered all the feedback. I've also updated the tests I added to use string interpolation (instead of concatenation) since that seems to be what most of the tests do. I've also added 2 test cases to the existing `AttributeArgumentSyntaxExtensionsTests` to cover the positive/negative `DateTimeOffset`-related tests
